src/responses/code_interpreter/service.rs

```diff
@@ -0,0 +1,342 @@
+    #[error("Failed to manage sandbox: {0}")]
+    Sandbox(String),
+    #[error("Failed to execute code: {0}")]
+    Execution(String),
+    #[error("Code execution timed out")]
+    Timeout,
+    #[error("Failed to acquire sandbox slot: {0}")]
+    Pool(String),
+}
+
+pub struct CodeInterpreterService {
+    docker: Docker,
+    settings: CodeInterpreterSettings,
+    sandboxes: Mutex<HashMap<String, SandboxEntry>>,
+    limiter: Arc<Semaphore>,
+}
+
+impl CodeInterpreterService {
+    pub async fn new(settings: CodeInterpreterSettings) -> Result<Self, CodeInterpreterError> {
+        let docker = Docker::connect_with_local_defaults()
+            .map_err(|err| CodeInterpreterError::DockerUnavailable(err.to_string()))?;
+
+        Ok(Self {
+            docker,
+            limiter: Arc::new(Semaphore::new(settings.max_sessions.max(1))),
+            sandboxes: Mutex::new(HashMap::new()),
+            settings,
+        })
+    }
+
+    pub async fn execute(
+        &self,
+        session_id: &str,
+        code: &str,
+    ) -> Result<ExecutionResult, CodeInterpreterError> {
+        self.reap_idle().await;
+
+        let container_id = self.ensure_sandbox(session_id).await?;
+        let result = self.exec_python(&container_id, code).await;
+        if result.is_err() {
+            self.teardown_session(session_id).await;
+        }
+        result
+    }
+
+    async fn ensure_sandbox(&self, session_id: &str) -> Result<String, CodeInterpreterError> {
+        if let Some(existing) = self
+            .sandboxes
+            .lock()
+            .await
+            .get_mut(session_id)
+            .map(|entry| {
+                entry.last_used = Instant::now();
+                entry.container_id.clone()
+            })
+        {
+            return Ok(existing);
+        }
+
+        let permit = self
+            .limiter
+            .clone()
+            .acquire_owned()
+            .await
+            .map_err(|err| CodeInterpreterError::Pool(err.to_string()))?;
+
+        let container_id = match self.start_container(session_id).await {
+            Ok(id) => id,
+            Err(err) => {
+                drop(permit);
+                return Err(err);
+            }
+        };
+
+        let mut guard = self.sandboxes.lock().await;
+        guard.insert(
+            session_id.to_string(),
+            SandboxEntry {
+                container_id: container_id.clone(),
+                last_used: Instant::now(),
+                _permit: permit,
+            },
+        );
+        Ok(container_id)
+    }
+
+    async fn start_container(&self, session_id: &str) -> Result<String, CodeInterpreterError> {
+        let sandbox_name = format!("nexus-ci-{session_id}");
+
+        let host_config = HostConfig {
+            memory: Some((self.settings.memory_limit_mb * 1024 * 1024) as i64),
+            nano_cpus: Some(((self.settings.cpu_percent * 1_000_000_000) / 100) as i64),
+            network_mode: Some("none".to_string()),
```

Comment:
`network_mode` sets to `"none"`, however, the code snippet between line 195 and 205 would invoke `pip install ...` command to install some packages?